File: sandbox-runner/patterns/batch-iterator/go/activities.go
package main

import (
	"context"
	"time"
)

func FetchPage(_ context.Context, offset int, pageSize int) ([]int, error) {
	end := offset + pageSize
	if end > TotalRecords {
		end = TotalRecords
	}
	page := make([]int, 0, end-offset)
	for i := offset; i < end; i++ {
		page = append(page, i)
	}
	return page, nil
}

func ProcessRecord(_ context.Context, recordID int) error {
	// Simulate processing work.
	time.Sleep(50 * time.Millisecond)
	return nil
}

File: sandbox-runner/patterns/batch-iterator/java/src/main/java/Activities.java
import io.temporal.activity.ActivityInterface;

import java.util.ArrayList;
import java.util.List;

@ActivityInterface
public interface Activities {
    List<Integer> fetchPage(int offset, int pageSize);
    void processRecord(int recordId);

    final class Impl implements Activities {
        @Override
        public List<Integer> fetchPage(int offset, int pageSize) {
            int end = Math.min(offset + pageSize, Shared.TOTAL_RECORDS);
            List<Integer> page = new ArrayList<>();
            for (int i = offset; i < end; i++) {
                page.add(i);
            }
            return page;
        }

        @Override
        public void processRecord(int recordId) {
            try {
                Thread.sleep(50);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }
}
